Auto extraction and splice

felipeji opened the pull request:
This PR adds routines to automatically extract and splice spectra from the reduced data cubes. It also includes a routine to splice the reduced data cubes so that a single spliced cube is produced from the two observed arms (when available). The main changes and additions are as follows:

New file `src/pywifes/extract_spec.py` with new functions that will perform the extraction of both blue and red spectra from the data cubes. We have defined different auxiliary functions in `extract_spec.py`, for example, to automatically detect the three brightest sources, or to plot the field showing the detected sources and the apertures considered for extracting the spectra.

New file `src/pywifes/splice.py`. The module includes functions based on previous scripts available within the team to splice the extracted spectra, i.e., to join the blue and red spectra into one single spectrum, interpolating the overlapping region. Based on the same function, we wrote a function to splice the entire cubes, so in addition to a blue and a red cube, we obtain a single cube covering the entire wavelength range observed.

We have also included a new function in `src/pywifes/data_classifier.py` for matching cubes coming from the same observation but different arms (blue/red).

New routines added in the `reduction_scripts/reduce_data.py` file to classify the resulting reduced spectra for pairing them accordingly and automatically extracting the source and splicing the spectra and the cube (when possible).

Finally, in the `reduction_scripts/reduce_data.py`, I have implemented a 'try' statement within the loop over the arms, allowing the reduction process to proceed even if data from only one arm is present.

timothyfrankdavies reviewed (changes requested):
A few suggested refactors and such, but the bulk of it looks good to me! Just some places you could dry up some repeated code, I think. A couple of changes related to checking arguments as well.

If any of the refactors are fiddly then don't worry about them, I think it will generally be quick changes & renames & such.

I haven't been through splice.py yet, but the rest I have. Will go through that one tomorrow.

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+
+            # SET SKIP ALREADY DONE FILES ?
+            skip_done = True
+            # skip_done=True
```
Yeah! Set as False now. I was testing this and left it as True.

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+                else:
+                    pass
+        except:
+            print("%s skipped" % arm)
```
That is nice! Done (and learnt!)

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
-    for blue_cube in blue_cubes:
-        shutil.move(blue_cube, destination_dir)
+
+    # Move reduced cubes to the data_products directory
+    for cubes_path in [red_cubes_path, blue_cubes_path]:
+        cubes = glob.glob(os.path.join(cubes_path, "*.cube.fits"))
+        for cube in cubes:
+            file = os.path.basename(cube)
+            shutil.move(cube, os.path.join(destination_dir,file))
```
Done!

felipeji commented on src/pywifes/data_classifier.py:
```diff
@@ -142,3 +142,53 @@ def classify(data_dir, naxis2_to_process = 0):
+
+
+
+
```
Done! Sorry, I keep making the same mistake :)

felipeji commented on src/pywifes/data_classifier.py:
```diff
@@ -142,3 +142,53 @@ def classify(data_dir, naxis2_to_process = 0):
+        return None
+    
+    # Define the pattern to match
+    pattern = r'UT.*?(?=\..*?\.fits)'
```
Your are right. I wrote it in that way just for setting the output name for the detection plot being created, but is so weak as it depends, required an specific structure on the name of the file to work. I just removed this function and change how the `file_name` is defined in the dictionary . Now the plots will be named after the base name with no extensions and no arm label in the name,  and replacing `.cube`  by  `_detection_plot.pdf`
e.g., for the cubes `OBK-701408-WiFeS-Blue-UT20240308T170030-5.cube.fits` , `OBK-701408-WiFeS-Red--UT20240308T170030-5.cube.fits ` you will have `OBK-701408-WiFeS-UT20240308T170030-5_detection_plot.pdf`

felipeji commented on src/pywifes/data_classifier.py:
```diff
@@ -142,3 +142,53 @@ def classify(data_dir, naxis2_to_process = 0):
+        ut_part = match.group(0)
+        return ut_part
+    else:
+        return None
```
I refactored this

felipeji commented on src/pywifes/data_classifier.py:
```diff
@@ -142,3 +142,53 @@ def classify(data_dir, naxis2_to_process = 0):
+    for paths in matched_paths:
+            matched_dict = {'Blue': None, 'Red': None, 'file_name': None}
+            for path in paths:
+                arm = fits.getheader(path)['ARM']
```
I implemented this plus the new approach to generate `'file_name'` as commented above.

felipeji commented on src/pywifes/data_classifier.py:
```diff
@@ -142,3 +142,53 @@ def classify(data_dir, naxis2_to_process = 0):
+
+
+
+
```
again... sorry :)

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+    sky_sub = extract_params["sky_sub"]
+    check_plot = extract_params["check_plot"]
+    border_width = extract_params["border_width"]
+    r_arcsec = extract_params["r_arcsec"]
```
I did it for being more clear, but you are right, the way, e.g. ` r_arcsec=extract_params["r_arcsec"]` is clear enough and we keep it more simple and short in the way you suggested.

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+    file_path = os.path.join(reduction_scripts_dir, f"params_extract_{obs_mode}.json")
+
+    with open(file_path, "r") as f:
+        extract_params = json.load(f)
```
Done! Yeah, I also think is better not to change to f"params_{config_name}.json"

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+    # Match cubes from the same observation based on DATE-OBS 
+    # ----------------------------------------------------------
+
+    matched_list = cube_matcher(reduced_cubes_paths)
```
Done! I changed to `matched_cubes` as now we also have the files names there not only paths.

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+                output = os.path.join(
+                    working_dir, destination_dir, spliced_spectrum_name
+                )
+                splice_spectra(blue_spec, red_spec, output)
```
We have tried this, but the extraction-splice workflow doesn't fit well within the reduction structure. For example, we have to apply the same extraction parameters to both arms, so these parameters should be repeated in both the blue and red arms in the JSON file, which makes things less clear for users. Also, implementing these routines within the step structure makes the code less clear. Extraction and splicing are technically not part of the reduction process. Therefore, one option here is to put all this functionality in a dedicated module.

timothyfrankdavies replied:
Yeah, a dedicated module would make sense, it will depend on what the users want. This whole script could be broken into separate modules with separate configs if needed. Then a much simpler 'reduce_data_and_splice.py' could run the whole process, for example.

felipeji replied:
We could split the code in modules in following PR, for example when we implement the structure the archive team wants .

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+            sky_sub=sky_sub,
+            check_plot=check_plot,
+            plot_output=plot_output
+        )
```
This routine detect the brighter sources, extract and the write the resulted spectra.  Having or not a plot of the field with the sources detected is optional. I have follow your recommendation and renamed as `auto_extract_and_save` for clarity

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+            pattern_red = os.path.join(
+                destination_dir,
+                os.path.basename(match_cubes["Red"]).replace("cube", "spec.ap*"),
+            )
```
Definitely nicer, done!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+    border_width=3,
+    sky_sub=False,
+    check_plot=False,
+    plot_output = 'detected_apertures_plot.pdf',
```
Done! `extract_spec.py`, `reduce_data.py` and `splice.py` went thought `black`.

felipeji commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +912,190 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+            for blue_spec, red_spec in zip(blue_specs, red_specs):
+                spliced_spectrum_name = os.path.basename(blue_spec).replace(
+                    "Blue", "Splice"
+                )
```
Done!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+    # Load in the data
+    
+    # Blue arm
+    if blue_cube_path is not None:
+        blue_sci, b_sci_hdr = fits.getdata(blue_cube_path, 0, header=True)
+        b_var, b_var_hdr = fits.getdata(blue_cube_path, 1, header=True)
+        blue_wcs = WCS(b_sci_hdr).celestial
+        binning_x, binning_y = np.int_(fits.getheader(blue_cube_path, 0)['CCDSUM'].split())
+
+    else:
+        blue_sci = None
+        blue_wcs = None
```
Sone! I'm aware here of having `binning_x` and `binning_y`  we will debug this soon checking if we may have the scenario whe we nee `red/blue_binning_x/y`

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+        red_sci, r_sci_hdr = fits.getdata(red_cube_path, 0, header=True)
+        r_var, r_var_hdr = fits.getdata(red_cube_path, 1, header=True)
+        red_wcs = WCS(r_sci_hdr).celestial
+        binning_x, binning_y = np.int_(fits.getheader(red_cube_path, 0)['CCDSUM'].split())
```
Yes! As commented above, we need to double check this sooner.

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+    -----
+    This function takes one or more data cubes and collapses them by computing the median
+    along the wavelength dimension. The resulting 2D array represents the
+    median image.
```
hahahaha! Sure you will have a lot of opportunities of revenge talking in the (proper) software engineering jargon!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+
+    # Automatic source detection in the collapsed cubes (red + blue)
+    npeaks = 3  # Number of peaks to be detected
+    # We use avoid the edges of size border_width also for the statistics
```
Yeah better. Thanks!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+
+    # Calculate pixel scale from binning
+    pixel_scale_x = binning_x  # arcsec/pix
+    pixel_scale_y = 1 / 2 * binning_y  # arcsec/pix
```
Done!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+        b = r_arcsec / pixel_scale_y
+
+        # Creates source apertures in the detected positions
+        source_aps = EllipticalAperture(positions, a=a, b=b)
```
done!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+            b_out = b * 4
+
+            sky_aps = EllipticalAnnulus(
+                positions, a_in=a_in, a_out=a_out, b_out=b_out, b_in=b_in
```
Yeahh!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+                write_1D_spec(red_flux, red_var, r_sci_hdr, r_var_hdr, red_output)
+
+        if check_plot:
+            plt.close("all")
```
This is something left from test when I was showing the plots interactively. Sorry!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+    source_ap : `photutils.aperture` object
+        Aperture object representing the source region to be extracted.
+
+    sky_ap : `photutils.aperture` object, optional
+        Aperture object representing the sky region to be extracted.
+        Default is None.
```
done!

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+    return np.array(fl), np.array(var)
+
+
+def writeFITS(sci_cube, var_cube, sci_header, var_header, output):
```
Removed.

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+                # Hide ticks and labels on both axes
+                ax1.set_xticks([])
+                ax1.set_yticks([])
+                ax1.axis("off")
```
Done! Less repetitive and more clear now

felipeji commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,504 @@
+                base_red_ouput = base_red.replace("cube.fits", "spec.ap%s.fits" % ap_index)
+                print("Saving red extracted spectra")
+                red_output = os.path.join(output_dir, base_red_ouput)
+                write_1D_spec(red_flux, red_var, r_sci_hdr, r_var_hdr, red_output)
```
Done! Everything putted inside `extract_and_save` function

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    # The following line of code results in a deprecation waring
+    # A sparce matrix in link list format
+    A = sp.lil_matrix((len(x), len(_x)))
+    for m in np.arange(len(x)):
```
Prefer using `i` instead of `m`. It gets very confusing very quickly when there's a different `i` floating around.

Maybe `fi` -> something like `new_fractional_pos`, `i` -> `new_pos`, `m` -> `i`?

felipeji replied:
Done!

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    # The following line of code results in a deprecation waring
+    # A sparce matrix in link list format
+    A = sp.lil_matrix((len(x), len(_x)))
+    for m in np.arange(len(x)):
```
I'd use `range` instead of `np.arange` for loops. `range` generates numbers as needed, whereas I think `np.arrange` will build an entire array of length len(x), then iterate through it.

felipeji replied:
Done1

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+            # Find L(x-i) for each of these
+            xk = np.pi * (fi[m] - np.array(ik) + EPSILON)
+            w = a * np.sin(xk) * np.sin(xk / a) / xk**2
+            A[m, ik] = w / w.sum()
```
`ik` -> `new_pos_range`? `xk` -> ..... unsure. `scaled_pos_range`??

`w` -> ...... `scaled_pos`?? `interpolated_results`???

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    return A
+
+
+# Unclear if needed
```
Make sure to remove the comment before merging.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    flux, fluxVar = join_spectra(blueSpec, redSpec)
+
+    if len(flux) == 1 :
+        print("No spectral overlap")
```
Fullstops.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+        return flux, fluxVar
+
+
+def splice_spectra(blue_spec_path, red_spec_path, output):
```
`output` -> `output_path`?

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+    # Join the spectra
+    print("Splicing spectra")
+    flux, fluxVar = join_spectra(blueSpec, redSpec)
```
As I commented on `join_cubes`, careful of it returning None, None. Similar fix to that comment.

felipeji replied:
Yeah, completely right. I updated as  `if flux == None`

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+        # Write out the results
+        # Use the header in red arm to start with
+        # Add additional blue CCD keywords as required - pending
```
Does this comment need an update? It ends with `pending`. If you'd like to track upcoming changes, you could make a github issue.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+            fluxvar_B = np.array((AB * diag_B * AB.T).sum(axis=1)).ravel()
+            fluxvar_R = np.array((AR * diag_R * AR.T).sum(axis=1)).ravel()
+
+            buffer = 10.0
```
This could be defined in SCREAMING_SNAKE_CASE, given it's a constant. Maybe above the 2 loops?

felipeji replied:
done!

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+
+# Unclear if needed
+class SingleSpec(object):
```
A class like this could be a nice fix for the complexity we're running into in `join_cubes`.

Maybe something similar, storing most of the same data as here, though maybe going further and adding functions to compute and store things like AB, flux_B, diag_B, given a particular row and column?? 

Might be easier than trying to rework the functions the other ways I've suggested.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+        self.wl = (
+            np.arange(self.header["NAXIS1"]) - self.header["CRPIX1"] + 1
+        ) * self.header["CDELT1"] + self.header["CRVAL1"]
+        # Temporary
```
Unsure what you mean by Temporary. Could probably remove the comment?

felipeji replied:
Done!

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+            diag_B = sp.dia_matrix(([blue_fluxvar], [0]), shape=[blueNAXIS1, blueNAXIS1])
+
+            fluxvar_B = np.array((AB * diag_B * AB.T).sum(axis=1)).ravel()
+            fluxvar_R = np.array((AR * diag_R * AR.T).sum(axis=1)).ravel()
```
You could group these a little differently I think.

I'd go for 
```py
            red_flux =  red_flux_cube[:, i, j]
            flux_R = np.array(AR * red_flux).ravel()

            red_fluxvar =  red_fluxvar_cube[:, i, j]
            diag_R = sp.dia_matrix(([red_fluxvar], [0]), shape=[redNAXIS1, redNAXIS1])
            fluxvar_R = np.array((AR * diag_R * AR.T).sum(axis=1)).ravel()

            blue_flux =  blue_flux_cube[:, i, j]
            flux_B = np.array(AB * blue_flux).ravel()

            # Two sparse diagonal matrices
            blue_fluxvar =  blue_fluxvar_cube[:, i, j]
            diag_B = sp.dia_matrix(([blue_fluxvar], [0]), shape=[blueNAXIS1, blueNAXIS1])
            fluxvar_B = np.array((AB * diag_B * AB.T).sum(axis=1)).ravel()
```
because in each of those, only the final line is used further down the function. 

You *could* then go and make each of those groups its own function call, but it's not really necessary.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+            overlap = np.where((wl < blue_maxWL - buffer) & (wl > red_minWL + buffer))
+            redOnly = np.where(wl >= blue_maxWL - buffer)
+
+            if red_minWL > blue_maxWL:
+                
+                return None, None
```
As noted in `splice_cubes`, you *might* want to reconsider the return values here, but also, you should do this check as soon as `red_minWL` and `blue_maxWL` are computed.

felipeji replied:
Yeah, done!

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+def splice_cubes(blue_path, red_path, output):
+    """
+    The main routine
```
This article is a stub :p

timothyfrankdavies reviewed (changes requested):
Went through `splice.py`. A few spots where you'll need to fix edge case handling, but the main thing were the sets of variables to handle through `join_cubes` and `join_spectra`. 

I didn't really figure out a good alternative, in `join_cubes` I suggested adding some extra functions to split things up, but in `join_spectra` I see you're using a class to keep track of the extra variables instead.

You don't necessarily have to do either, a bit of regrouping and renames could help keep things readable.

I'd suggest taking a greedy strategy: try the easiest approach, then just keep trying the next easiest approach until you get something you're happy with :p

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    flux, fluxVar = join_cubes(blue_path, red_path)
+
+    if len(flux) == 1 :
+        print("No spectral overlap")
```
Done!

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+
+
+def splice_cubes(blue_path, red_path, output):
```
done!

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    print("Splicing cubes")
+    flux, fluxVar = join_cubes(blue_path, red_path)
+
+    if len(flux) == 1 :
```
Yeah, agreed! Also I refactor `join_cubes` to set   `if red_minWL > blue_maxWL earlier` earlier and save time in case cubes doesn't overlap.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    red_flux_cube = fits.getdata(red_path, 0, header=False)
+    red_fluxvar_cube = fits.getdata(red_path, 1, header=False)
+
+    redCRVAL1 = fits.getheader(red_path)["CRVAL3"]
+    redCDELT1 = fits.getheader(red_path)["CDELT3"]
+    redCRPIX1 = fits.getheader(red_path)["CRPIX3"]
+    redNAXIS1 = fits.getheader(red_path)["NAXIS3"]
+    red_wl = (np.arange(redNAXIS1) - redCRPIX1 + 1) * redCDELT1 + redCRVAL1
+    red_minWL = min(red_wl)
+    red_maxWL = max(red_wl)
```
I have renamed some of the variables. I kept headers acronyms as they are pretty well know standards in astronomy.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    blueCRPIX1 = fits.getheader(blue_path)["CRPIX3"]
+    blueNAXIS1 = fits.getheader(blue_path)["NAXIS3"]
+    blue_wl = (np.arange(blueNAXIS1) - blueCRPIX1 + 1) * blueCDELT1 + blueCRVAL1
+    blue_minWL = min(blue_wl)
```
yeah!

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    return abs((dx.max() - dx.min()) / np.median(dx)) < 1e-3
+
+
+def _A_lanczos(_x, x, a=2, missing=np.nan):
```
I dos some refactor here following your advise.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    red_maxWL = max(red_wl)
+
+    # Read blue data and metadata
+    blue_flux_cube = fits.getdata(blue_path, 0, header=False)
```
Yeah, that's more clear!

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    blueNAXIS1 = fits.getheader(blue_path)["NAXIS3"]
+    blue_wl = (np.arange(blueNAXIS1) - blueCRPIX1 + 1) * blueCDELT1 + blueCRVAL1
+    blue_minWL = min(blue_wl)
+    blue_maxWL = max(blue_wl)
```
Instead of a function, I read the header into a variable and then just take the corresponding keys. I think that, even when is a bit repetitive, we this structure and the new variable name you suggested we have it now pretty clear.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    # set to the blue spectrum. However the code is robust to this.
+
+    AB = _A_lanczos(blue_wl, wl, 3).tocsr()
+    AR = _A_lanczos(red_wl, wl, 3).tocsr()
```
Legacy code here, Simba :) .... I'm totally see your point but I'm not confident enough changing variables name here without going in detail and do some extra test.  I have change some of the variables names, but I prefer to back on this with more detail on futures PRs if you are ok with that

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+        hdr_fluxvar["CDELT3"] = blue_header["CDELT3"]
+        hdr_fluxvar["CTYPE3"] = "wavelength"
+        hdr_fluxvar["CUNIT3"] = "angstrom"
+        hdr_fluxvar["BUNIT"] = "(count / Angstrom)^2"
```
This is the result of adapting an existing function doing the same considering also the standard form of headers in 1D spectra.  Anyway I will double check this with Chris.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    i = np.array([int(f) for f in fi])
+    # Write down the matrix for Lanczos interpolation
+    EPSILON = np.finfo(np.double).eps
+    # The following line of code results in a deprecation waring
```
That comment was there from previous version, sorry for not having removed it! I have no deprecation warning prompted so far. I'll keep on eye on it.

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+
+
+    nwl = int((redCRVAL1 + redCDELT1 * redNAXIS1 - blueCRVAL1) / blueCDELT1)
+    wl = (np.arange(nwl) - blueCRPIX1 + 1) * blueCDELT1 + blueCRVAL1
```
Done!

felipeji commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,330 @@
+    # set to the blue spectrum. However the code is robust to this.
+
+    AB = _A_lanczos(blue_wl, wl, 3).tocsr()
+    AR = _A_lanczos(red_wl, wl, 3).tocsr()
```
Yeah, your are right. The thing is I didn't go very deep into `_A_lanczos`. I limited to replicated what it was already wrote for splicing spectra but  running over all the index in the spatial dimensions of the cube. I'll leave it as it is now. We could back for refactoring  `_A_lanczos` a bit and the things related with it.

felipeji commented:
I have implemented most of your (useful!) comments.  I have also run some tests for checking everything still running properly. On one of those test I found a error when more than one aperture is detected. The error come from handling header instead of a copy of it, so the changes in the header produced when operating for one aperture remains in the header variable and trigger and error when try to change the header again for the next iteration (aperture).  I fixed by doing  (`splice.py`, line 192):

`hdulist[0].header = redSpec.header`  ->  `hdulist[0].header = redSpec.header.copy()`

I tried to implement most of the changes in the (now called) `a_lanczos` function. As I commented along the review, I feel not too confident on change it further without testing it carefully, so prefer to do that in a future refactoring.

timothyfrankdavies commented:
> I have implemented most of your (useful!) comments. I have also run some tests for checking everything still running properly. On

Good work! Hadn't realised the headers were getting overwritten and reused and such. Leaving some refactors to later makes sense to me. Provided the code's in a state where it's easy enough to follow, they can even wait until the code is hit by other PRs making functional changes.

I'll give your updates a look today, then should be good to go!

timothyfrankdavies commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +936,178 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+                else:
+                    pass
+        except Exception as exc:
+            print("{arm} skipped, as an error occurred during processing: '{exc}'")
```
A format string needs an 'f' at the front, so `print(f"{arm} skipped, as an error occurred during processing: '{exc}'.")`, added a fullstop too.

felipeji replied:
oh yes, sorry for the typos.

timothyfrankdavies commented on reduction_scripts/reduce_data.py:
```diff
@@ -896,99 +936,178 @@ def run_save_3dcube(metadata, prev_suffix, curr_suffix, **args):
+            blue_cube_path,
+            red_cube_path,
+            destination_dir,
+            r_arcsec=extract_params["r_arcsec"],
+            border_width=extract_params["border_width"],
+            sky_sub=extract_params["sky_sub"],
+            check_plot=extract_params["check_plot"],
```
Nice having the parameter names & headers line up, easy to verify!

timothyfrankdavies commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,516 @@
+                sky_ap = None
+            ap_index = index + 1
+
+            # Extracting blue cube
+            extract_and_save(
+                blue_cube_path,
+                blue_sci,
+                blue_var,
+                source_ap,
+                ap_index,
+                sky_ap,
+                output_dir,
+                blue_sci_hdr,
+                blue_var_hdr,
+            )
+
+            # Extracting red cube
+            extract_and_save(
+                red_cube_path,
+                red_sci,
+                red_var,
+                source_ap,
+                ap_index,
+                sky_ap,
+                output_dir,
+                red_sci_hdr,
+                red_var_hdr,
+            )
```
Much easier to verify they're doing the same thing when you just have to compare the arguments, ty!

timothyfrankdavies commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,516 @@
+        sci, sci_hdr = fits.getdata(cube_path, 0, header=True)
+        var, var_hdr = fits.getdata(cube_path, 1, header=True)
+        wcs = WCS(sci_hdr).celestial
+        binning_x, binning_y = np.int_(fits.getheader(cube_path, 0)["CCDSUM"].split())
```
`fits.getheader(cube_path, 0)` could just be `sci_hdr`.

felipeji replied:
Done!

timothyfrankdavies commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,516 @@
+    blue_wcs = blue_cube_data["wcs"]
+    if blue_cube_data["sci"] is not None:
+        binning_x = blue_cube_data["binning_x"]
+        binning_y = blue_cube_data["binning_y"]
```
This is better, though you could do the same as reduce_data.py and use the dictionary lookup inline instead of making the intermediate variables. But, I'd say not worth changing, huge improvement as-is!

felipeji replied:
We still need to check with Chris and Ian if we need to consider different binning configuration for red and blue arm.  I'l keep as it is now but I'll used dictionary lookup then.

timothyfrankdavies commented on src/pywifes/extract_spec.py:
```diff
@@ -0,0 +1,516 @@
+
+    # Update WCS headers to match the 1D spectrum header
+    sci_cube_header_copy = sci_cube_header.copy()
+    var_cube_header_copy = var_cube_header.copy()
```
Ahhh right, yes makes sense that you'd need a copy, and it makes sense intuitively that you wouldn't expect the header you pass in to be overwritten.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,360 @@
+
+def join_cubes(blue_path, red_path):
+
+    # Read red data and metadata
+    red_flux_cube, blue_header = fits.getdata(red_path, 0, header=True)
+    red_fluxvar_cube = fits.getdata(red_path, 1, header=False)
+    red_CRVAL1 = blue_header["CRVAL3"]
+    red_CDELT1 = blue_header["CDELT3"]
+    red_CRPIX1 = blue_header["CRPIX3"]
+    red_NAXIS1 = blue_header["NAXIS3"]
+    red_wavelength = (np.arange(red_NAXIS1) - red_CRPIX1 + 1) * red_CDELT1 + red_CRVAL1
+    red_min_wavelength = min(red_wavelength)
+
+    # Read blue data and metadata
+    blue_flux_cube, red_header = fits.getdata(blue_path, 0, header=True)
+    blue_fluxvar_cube = fits.getdata(blue_path, 1, header=False)
+    blue_CRVAL1 = red_header["CRVAL3"]
+    blue_CDELT1 = red_header["CDELT3"]
+    blue_CRPIX1 = red_header["CRPIX3"]
+    blue_NAXIS1 = red_header["NAXIS3"]
+    blue_wavelength = (
+        np.arange(blue_NAXIS1) - blue_CRPIX1 + 1
+    ) * blue_CDELT1 + blue_CRVAL1
+    blue_max_wavelength = max(blue_wavelength)
```
I think the names `red_header` and `blue_header` got switched somehow. The actual logic is fine, just wrong variable names I think.

felipeji replied:
Yes! my bad. Thanks for noting it!

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,360 @@
+        return None, None
+
+    # Cubes overlap in wavelength
+    else:
```
If a previous `if` results in a function returning/exiting/etc, then you don't *strictly* need an else, since the condition *must* be false if you've reached that code.

That said, it can be hard to keep track of how many conditions are effectively false if there isn't the corresponding code indentation, so opinions vary as to which one is more readable.

In this case, the code below is a bit indented already, and it's all group up pretty neatly, so I'd lean towards removing the 'else'. Either's fine though.

timothyfrankdavies commented on src/pywifes/splice.py:
```diff
@@ -0,0 +1,360 @@
+
+    else:
+
+        # Write out the results
+        # Use the header in red arm to start with
+        # Add additional blue CCD keywords as required
```
Fullstops could be nice, hard to tell if it's 3 separate comments, or 1 multiline comment.

felipeji replied:
Yeah, I put them all in a single line.

timothyfrankdavies reviewed (approved):
A few minor comments, one variable name mixup, but nothing that really needs fixing.

I'm happy to approve, probably worth fixing the mixup but I don't think I'd need to re-review the change. Nice work!
